cmd/compose/pull.go

```diff
@@ -83,6 +85,13 @@ func runPull(ctx context.Context, dockerCli command.Cli, backend api.Service, op
 		}
 	}
 
+	if opts.policy != "" {
+		for i, service := range project.Services {
+			service.PullPolicy = opts.policy
```

Comment:
I think this should ignore services that either have `pull_policy: build` already set in YAML or have a `<svc>.build` section and no `<svc>.image` section.

e.g. imagine this project
```yaml
services:
  explicit-build:
    image: registry.example.com/myservice
    pull_policy: build
    build: .
  no-image:
    build: .
  nginx:
    image: nginx
```

Without args:
```
❯ docker compose pull
[+] Pulling 3/3
 ✔ explicit-build Skipped                                                                                                                             0.0s
 ✔ no-image Skipped - No image to be pulled                                                                                                           0.0s
 ✔ nginx Pulled
```

With `--policy=missing` (or `--policy=always`):
```
❯ ~/dev/compose/bin/build/docker-compose pull --policy=missing
[+] Pulling 3/3
 ✔ no-image Skipped - No image to be pulled                                                                                                           0.0s
 ✔ nginx Pulled                                                                                                                                       0.5s
 ! explicit-build Warning                                                                                                                             0.2s
WARNING: Some service image(s) must be built from source by running:
    docker compose build explicit-build
1 error occurred:
	* Error response from daemon: Get "https://registry.example.com/v2/": dialing registry.example.com:443 with direct connection: resolving host registry.example.com: lookup registry.example.com: no such host
```

So it is properly skipping the one without `<svc>.image`, but is unusable if you have any `pull_policy: build` services that do specify an image name.

Reply:
Agree about service without an `image` MUST be excluded
disagree on `pull_policy: build`: we should only require a local build if image is not available on registry under configured image